### PR TITLE
Extended Loadout Framework

### PR DESCRIPTION
**Pull request by BrettMayson**
**When merged this pull request will:**
- Enable mods to store extended info about a player's loadout
- Superceed #1270, this system can also be used to filter the loadouts, to remove radio IDs as an example

<details><summary>ACE Earplugs Example</summary>

```sqf
["CBA_loadoutSet", {
    params ["_unit", "_loadout", "_extradata"];
    private _hasEarplugs = _extradata getOrDefault ["ace_earplugs", false];
    _unit setVariable ["ACE_hasEarPlugsIn", _hasEarplugs, true];
    [[true]] remoteExec ["ace_hearing_fnc_updateVolume", _unit];
}] call CBA_fnc_addEventHandler;

["CBA_loadoutGet", {
    params ["_unit", "_loadout", "_extradata"];
    if (_unit getVariable ["ACE_hasEarPlugsin", false]) then {
        _extradata set ["ace_earplugs", true]
    };
}] call CBA_fnc_addEventHandler;
```

</details>

<details><summary>ACRE Flagged Radio example</summary>

```sqf
["CBA_loadoutGet", {
    params ["", "_loadout", ""];
    if ((_loadout select 9) select 2 == "ItemRadioAcreFlagged") then {
        (_loadout select 9) set [2, ""];
    };
    
    // Set ACRE base classes
    private _replaceRadioAcre = {
        params ["_item"];
        // Replace only if string (array can be eg. weapon inside container) and an ACRE radio
        if (!(_item isEqualType []) && {[_item] call acre_api_fnc_isRadio}) then {
            _this set [0, [_item] call acre_api_fnc_getBaseRadio];
        };
    };
    if !((_loadout select 3) isEqualTo []) then {
        {_x call _replaceRadioAcre} forEach ((_loadout select 3) select 1); // Uniform items
    };
    if !((_loadout select 4) isEqualTo []) then {
        {_x call _replaceRadioAcre} forEach ((_loadout select 4) select 1); // Vest items
    };
    if !((_loadout select 5) isEqualTo []) then {
        {_x call _replaceRadioAcre} forEach ((_loadout select 5) select 1); // Backpack items
    };
}] call CBA_fnc_addEventHandler;
```

</details>

Usage is very similar to the vanilla `getUnitLoadout` when applied to units

```sqf
private _loadout = [player] call CBA_fnc_getLoadout;
[player, _loadout, true] call CBA_fnc_setLoadout
```

### CODE REVIEW

**Comment by commy2** [comment]
Very clean code.

Why do we need the whole LoadoutSet framework though? Couldn't we just raise and listen to "vanilla" CBA events?

**Comment by BrettMayson** [comment]
We'd still need to store the defaults for those somewhere, and I don't see any advantage of events over a dedicated functiom

File: addons/loadout/fnc_addLoadoutSetHandler.sqf [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        }] call CBA_fnc_addLoadoutSetHandler;
+    (end)
+Author:
+    Brett Mayson
+---------------------------------------------------------------------------- */
+
+params [
+    ["_id", "", [""]],
+    ["_function", {}, [{}]],
+    "_default"
```
Comment (commy2) [inline addons/loadout/fnc_addLoadoutSetHandler.sqf:33]: This is undocumented.

**Comment by commy2** [comment]
> We'd still need to store the defaults for those somewhere, and I don't see any advantage of events over a dedicated function

Maybe make it a wrapper that uses CBA events?

I would really like to avoid duplicating the events logic yet again.

File: addons/loadout/fnc_setLoadout.sqf [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    private _id = _x;
+    {
+        _x params ["_function", "_default"];
+        if (_id in _extendedInfo) then {
+            [_unit, _extendedInfo get _id] call _function;
+        } else {
+            [_unit, _default] call _function;
+        };
+    } forEach (GVAR(setHandlers) get _id);
+} forEach (keys GVAR(setHandlers));
```
Comment (mharis001) [inline addons/loadout/fnc_setLoadout.sqf:42]: Can just `forEach` the `HashMap` directly. Value is in `_y`.

File: addons/loadout/fnc_setLoadout.sqf [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+{
+    private _id = _x;
+    {
+        _x params ["_function", "_default"];
+        if (_id in _extendedInfo) then {
+            [_unit, _extendedInfo get _id] call _function;
+        } else {
+            [_unit, _default] call _function;
+        };
```
Comment (mharis001) [inline addons/loadout/fnc_setLoadout.sqf:40]: Can use `getOrDefault` to simplify this code.

File: addons/loadout/fnc_setLoadout.sqf [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    Brett Mayson
+---------------------------------------------------------------------------- */
+
+params [
+    ["_unit", objNull, [objNull]],
+    ["_loadout", [], [[]]],
+    ["_fullMagazines", false, [false]]
+];
+
+if (_unit isEqualTo objNull) exitWith {};
```
Comment (mharis001) [inline addons/loadout/fnc_setLoadout.sqf:26]: `isNull`.

**Comment by BrettMayson** [comment]
Any updates on this?

**Comment by BrettMayson** [comment]
Anyone want to take on modifying this to commy's requirements? I do not have the time.

**Comment by BrettMayson** [comment]
@commy2 Updated as discussed on Slack

File: addons/loadout/config.cpp [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#include "script_component.hpp"
+
+class CfgPatches {
+    class ADDON {
+        name = CSTRING(component);
```
Comment (PabstMirror) [inline addons/loadout/config.cpp:5]:
```suggestion
        name = "Community Base Addons - Loadout Framework"
```

File: addons/loadout/fnc_setLoadout.sqf [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+if (isNull _unit) exitWith {};
+
+// A regular loadout array was passed in
+if (count _loadout == 10) exitWith {
+    _unit setUnitLoadout [_loadout, _fullMagazines];
+};
+
+_unit setUnitLoadout [_loadout select 0, _fullMagazines];
+
+private _extendedInfo = createHashMapFromArray (_loadout select 1);
```
Comment (PabstMirror) [inline addons/loadout/fnc_setLoadout.sqf:35]: getLoadout returns hash, so I think we'd want something like this to handle either hash or arrays?
```suggestion
private _extendedInfo = _loadout select 1;
if (_extendedInfo isEqualType []) then { _extendedInfo = createHashMapFromArray _extendedInfo; };
```

File: addons/loadout/config.cpp [hunk 7]
```diff
@@ -0,0 +1,17 @@
+#include "script_component.hpp"
+
+class CfgPatches {
+    class ADDON {
+        name = "Community Base Addons - Loadout Framework"
```
Comment (PabstMirror) [inline addons/loadout/config.cpp:5]:
```suggestion
        name = "Community Base Addons - Loadout Framework";
```

File: addons/loadout/fnc_setLoadout.sqf [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+if (isNull _unit) exitWith {};
+
+// A regular loadout array was passed in
+if (count _loadout == 10) exitWith {
+    _unit setUnitLoadout [_loadout, _fullMagazines];
+};
+
+_unit setUnitLoadout [_loadout select 0, _fullMagazines];
+
+private _extendedInfo = _loadout select 1;
```
Comment (mharis001) [inline addons/loadout/fnc_setLoadout.sqf:35]: `params` instead of `select`.

**Review by PabstMirror — APPROVED** [review]

**Pull request merged**